whylog/assistant/regex_assistant/regex_object.py

```diff
@@ -8,69 +10,80 @@ class RegexObject(object):
     Verifies updates.
-    Not responsible for guessing where are groups in regex
+
+    :param line_text: line text (raw string)
+    :param param_groups: represents params from text (catched by regex groups)
+    :param regex: regex matching to line_text
+    :param guessed_pattern_objects: keeps guessed regexes that match to line_text
+    :type param_groups: dict[int, ParamGroup]
+    :type guessed_pattern_objects: dict[int, PatternObject]
     """
 
     def __init__(self, line_object):
-        """
-        :param FrontInput line_object: object that represents line
-        :param line_text: line text (raw string)
-        :param regex: regex corresponding to line_text
-        :param [Span] group_spans: represents regex groups, not overlapping each other
-        """
-        self.line = line_object
         self.line_text = line_object.line_content
-        self.group_spans = SpanList()
-        self.regex = self._update_regex()
+        self.param_groups = dict()
 
-    def _hole_spans(self):
-        """
-        Returns spans between group_spans in line_text
+        self.regex = None
+        self.update_by_regex(create_obvious_regex(self.line_text))
 
-        I.e having line_text of length 50 and group_spans corresponding to intervals: (1, 21), (30, 41),
-        returns complementary spans corresponding to intervals (0, 1), (21, 30), (41, 50)
-        """
-        compl_intervals = self.group_spans.complementary_intervals(0, len(self.line_text))
-        compl_spans = SpanList.from_ranges(compl_intervals, pattern_creator=create_obvious_regex)
-        return compl_spans
+        self.guessed_pattern_objects = dict()
+        self._guess_regexes()
+
+    def convert_to_pattern_object(self):
+        return PatternObject(self.line_text, self.regex, self.param_groups)
 
-    def _update_regex(self, force=False):
+    def update_by_regex(self, new_regex):
         """
-        Builds regex from spans
+        Assigns self.regex to new_regex.
+
+        Throws NotMatchingRegexError if new_regex doesn't match self.line_text
+        If needed, improves new_regex: new_regex = '^' + new_regex + '$' (regex must match whole line)
+        Updates self.param_groups so that they correspond to new_regex groups
         """
 
-        hole_spans = self._hole_spans()
+        groups = verify_regex(new_regex, self.line_text)
 
-        line_spans = (hole_spans + self.group_spans).sort_by_start()
+        if not new_regex[0] == '^':
+            new_regex = '^' + new_regex
+        if not new_regex[-1] == '$':
+            new_regex += '$'
 
-        regex = r""
-        for span in line_spans:
-            span.update_pattern(self.line_text, force)
-            span_pattern = span.pattern
-            if span.is_param:
-                span_pattern = "(" + span_pattern + ")"
-            regex += span_pattern
-        regex = "^" + regex + "$"
-        return regex
+        default_converter = DataType.STRING
+        self.param_groups = dict(
```

Comment:
`self.param_groups = {key + 1: ParamGroup(groups[key], default_converter) for key in range(len(groups))}`
much simpler and shorter.